sshd-common/src/main/java/org/apache/sshd/common/config/keys/OpenSshCertificate.java

```diff
@@ -52,26 +57,34 @@
     Collection<String> getPrincipals();
 
     /**
-     * Retrieves the time in number of seconds since the {@link java.time.Instant#EPOCH} at which this certificate
-     * becomes or became valid.
-     *
-     * @return the number of seconds since the Instant.EPOCH <em>as an unsigned 64bit value</em>
-     * @see    {{@link #isValidNow(OpenSshCertificate)}
+     * When null, implies forever
      */
-    long getValidAfter();
+    Instant getValidAfter();
+
+    default long getValidAfterEpochSeconds() {
+        if (getValidAfter() == null) {
+            return VALID_AFTER_FOREVER_EPOCH;
+        }
+        return getValidAfter().getEpochSecond();
```

Comment:
I don't understand why you handle validAfter == 0 specially. That would be Instant.EPOCH.

I think you can handle all values in the range 0 .. Long.MAX_VALUE normally. If it's greater, the signed value will appear as < 0. Here I think handling the ~0L case ("forever") as null is fine; all others I'd map to Long.MAX_VALUE.

Reply:
My observation was that for the `validAfter` field, a certificate has a uint64 encoded as `0` in the case of a "forever" certificate, so my intention was to treat this as a special case in two ways:

1. during decoding (in `OpenSSHCertPublicKeyParser`), when it decodes the bytes as a uint64 `0`, leave the `Instant` as null (to signify forever)
2. On `OpenSshCertificate.getValidAfterEpochSeconds` (simply a helper), apply this same logic in reverse, if the value is null, return the magic "forever" epoch that should be encoded

The intention would be that the business code a developer writes only needs to interact with the `Instant` (null for forever, otherwise set a value), and `OpenSshCertificate.getValidAfterEpochSeconds` and `OpenSshCertificate.getValidBeforeEpochSeconds` are simply used when encoding the bytes of the certificate into a `Buffer`

Reply:
Can't validAfter be legally zero if I specify the epoch as from date? I don't see the zero value as special.

But the conversion really has to handle 64bit validLong > Long.MAX_VALUE, also for validAfter. Effectively that would be a certificate that is never valid. But `System.out.println(Instant.ofEpochSecond(-2L))` (0xffff_ffff_ffff_fffeL) yields `1969-12-31T23:59:58Z`.

Reply:
I don't think it can be legally `0` and mean "forever"

I just re-read the man page for `ssh-keygen`'s `validity_interval` and you can use the magic values `always` for the `validAfter` and `forever` for `validBefore`

But, when encoded, there is no special flag to indicate this, they are still just epoch (seconds) uint64's, where `always` (`validAfter`) is encoded as `0` and `forever` (`validBefore`) is uint64 max value

I manually verified this by generating certs and looking at the encoded bytes, and you can also see it like this, where generating a certificate using `-V 'always:20220101' versus `-V '19790101:20220101` yield identical values.

If you inspect the cert using `ssh-keygen -L -f [file]` when using a cert generated with `-V '19790101:20220101`, as you would now expect, it prints a human-friendly output of what "always" would have shown.

What's more, is if you move the date back before UTC 1970, ssh-keygen still encodes it as `0` (which I guess they have to, since it's a uint64) 

... so maybe when they defined the spec they assumed (probably rightly so) that no one would be trying to enforce a `validAfter` time on a server that was pretending to run pre 1970?

So, it seems to me, like they do treat `0` and max uint64 as a magic value.  However, it is also entirely possible this magic behavior is just for the human-friendly output of the CLI tools...

I'm happy to change this to behave how you think it should, but I'm not really sure what that should be

Reply:
My point was not that someone would try to specify pre-1970 dates. The point was: if a certificate has a validAfter/Before >= 0x8000_0000_0000_0000 (unsigned long), that's a time some 300 billion years in the future. If this is passed to Instant.ofEpochSecond(), you get an exception. Some other large values (>= 0xff8f_e310_1464_1400 in unsigned long; Instant.MIN_SECOND) are even much farther in the future but will be taken as negative values and yield pre-1970 instants in Java instead.

Exactly:
> -V 'always:20220101' versus -V '19700101:20220101' yield identical values.

So zero is not special; but OpenSSH has a special token "always" for it.

If the 64bit value is >= Instant.MAX_SECOND (0x70_1cd2_fa95_78ffL) or, interpreted as signed, is <= Instant.MIN_SECOND, converting the 64bit value to an Instant is going to be problematic.

So... full round-tripping with Instant is not possible. If you want full round-tripping capabilities, long and being careful as a developer is the only way.

For _setting_ validAfter/validBefore, Instant is the right interface if there is a validation that the `value == null || value.compareTo(Instant.EPOCH) >= 0`, serializing null as 0xffff_ffff_ffff_ffffL and Instant.getEpochSecond() otherwise. For decoding the timestamp of a certificate, one would have to handle the special cases:

- value in the range [0 .. Instant.MAX_SECOND]: Instant.ofEpochSecond(value)
- 0xffff_ffff_ffff_ffffL: null
- any other: Instant.MAX

Instant.MAX is still nearly a billion years in the future, which is good enough for all practical purposes.

Reply:
Would you prefer to continue to use Instant and layer on these logical constraints, or go back to long's and apply fewer logical constraints (probably just a >= 0 check to ensure it's treated as a signed value)?

I would rather continue to use Instant, even with the constraints, it's still fine for all practical purposes

Reply:
I think it boils down to "do we need full round-tripping?"

I lack experience in this area. I like interfaces with Instant better, too, but I'm worried about not being able to represent the full range of possible values. Couldn't this become a problem for verifying a certificate signature? At least internally I think we'll have to keep the raw 64bit value for this case. Also, if we ever want to read a received certificate, pass it around as OpenSshCertificate object, and then write it out again. We mustn't change the bits of the validAfter/validBefore fields, that would invalidate the signature.

For creating OpenSshCertificateSignRequests, Instant is OK (with validation for null or >= epoch).

Reply:
All very good points, I'll make adjustments to use longs as raw storage and some optional helper(s) for Instant (probably only useful on the setter, based on our discussion)

Reply:
I restored the raw storage to `long` and experimented with a few ideas to try and give a better developer UX, but, the best I could come up with was https://github.com/apache/mina-sshd/pull/198/commits/ffaa35ae566aa04a79a3bf94625af4d970756a55#diff-0b82051894f817aa54f0e169b2bb96d0fce8a989838e0dd4574ca7ab395d5f9eR199-R201

Which is just to have a helper that can set the values as an `Instant` but not before `Instant.EPOCH`